torch/autograd/gradcheck.py

```diff
@@ -249,13 +249,13 @@ def fail_test(msg):
         return False
 
     tupled_inputs = _as_tuple(inputs)
-    if any(t.is_sparse for t in tupled_inputs if isinstance(t, torch.Tensor)) and not check_sparse_nnz:
+    if not check_sparse_nnz and any(t.is_sparse for t in tupled_inputs if isinstance(t, torch.Tensor)):
         return fail_test('gradcheck expects all tensor inputs are dense when check_sparse_nnz is set to False.')
 
     # Make sure that gradients are saved for at least one input
     any_input_requiring_grad = False
     for idx, inp in enumerate(tupled_inputs):
-        if isinstance(inp, torch.Tensor) and inp.requires_grad:
+        if isinstance(inp, torch.Tensor) or hasattr(inp, "__torch_function__") and inp.requires_grad:
```

Comment:
Maybe add a function in this file: `is_tensor_like(inp)`  as you do this check quite a few times.

Reply:
I have a feeling such a function could be more generally useful. If so, where would be the best place to put it?

Reply:
`torch.overrides`?